sql/000002.sql

```diff
@@ -0,0 +1,183 @@
+);
+create function graphile_worker.add_job(
+  identifier text,
+  payload json = null,
+  queue_name text = null,
+  run_at timestamptz = null,
+  max_attempts int = null,
+  job_key text = null
+) returns graphile_worker.jobs as $$
+declare
+  v_job graphile_worker.jobs;
+begin
+  if job_key is null then
+    insert into graphile_worker.jobs(task_identifier, payload, queue_name, run_at, max_attempts)
+      values(
+        identifier,
+        coalesce(payload, '{}'),
+        coalesce(queue_name, public.gen_random_uuid()::text),
+        coalesce(run_at, now()),
+        coalesce(max_attempts, 25)
+      )
+      returning *
+      into v_job;
+  else
+    insert into graphile_worker.jobs (task_identifier, payload, queue_name, run_at, max_attempts, key)
+      values(
+        identifier,
+        coalesce(payload, '{}'),
+        coalesce(queue_name, public.gen_random_uuid()::text),
+        coalesce(run_at, now()),
+        coalesce(max_attempts, 25),
+        job_key
+      )
+      on conflict (key) do update set
+        -- update job details if provided, otherwise maintain existing value
+        -- note we can't use excluded here, or the value will always be updated,
+        -- due to the coalesce statements in values ( ... ) above
+        task_identifier=coalesce(add_job.identifier, jobs.task_identifier),
+        payload=coalesce(add_job.payload, jobs.payload),
+        queue_name=coalesce(add_job.queue_name, jobs.queue_name),
+        max_attempts=coalesce(add_job.max_attempts, jobs.max_attempts),
+
+        -- update run_at if argument is provided. If not, and there has been an
+        -- error, assume run_at reflects the error retry backoff, so reset it.
+        -- If no errors, maintain existing value by default
+        run_at=coalesce(
+          add_job.run_at,
+          (case when jobs.attempts > 0
+            then now()
+            else jobs.run_at
+          end)
+        ),
+
+        -- always reset error/retry state
+        attempts=0,
+        last_error=null
+      where jobs.locked_at is null
+      returning *
+      into v_job;
+
+    -- if the returned id is null, assume job is already locked for processing
+    -- and couldn't be updated
+    if v_job.id is null then
+      -- remove existing key to allow a new one to be inserted, and prevent any
+      -- subsequent retries by bumping attempts to the max allowed
+      update graphile_worker.jobs set
+        key = null,
+        attempts = jobs.max_attempts
+      where key = job_key;
+
+      -- insert the new job. Assume no conflicts due to the update above
+      insert into graphile_worker.jobs(task_identifier, payload, queue_name, run_at, max_attempts, key)
+        values(
+          identifier,
+          coalesce(payload, '{}'),
+          coalesce(queue_name, public.gen_random_uuid()::text),
+          coalesce(run_at, now()),
+          coalesce(max_attempts, 25),
+          job_key
+        )
+        returning *
+        into v_job;
```

Comment:
Since this is the exact code as at the start, for ease of maintenance / DRY can we flip the logic? Something like

```sql
if job_key is not null then
  insert ... on conflict ... returning * into v_job;
  if not (v_job is null) then -- IMPORTANT: this is not the same as `if v_job is not null`
    return v_job;
  else
    update jobs set key = null...;
    -- We'll re-attempt insert without ON CONFLICT below
  end if;
end if;

insert into ... values ... returning * into v_job;
return v_job;
```